### ISSUE TITLE
API versioning with blueprints causes colission

### ISSUE THREAD

**Issue by alejoar** [pallets/flask#1774]
The time has come where I need to release a new version for my api and still support the previous one.

I have an app structure that looks like this:
```
+-- app/
    +-- api_2_0/
        +-- __init__.py
        (...)
    +-- api_2_1/
        +-- __init__.py
        (...)
   +-- __init__.py
```

Both are blueprints which I initialize this way in my `__init__.py` in the `create_app` method (I'm using the [app factory](http://flask.pocoo.org/docs/0.10/patterns/appfactories/) method):
```
    from .api_2_0 import api as api_2_0_blueprint
    app.register_blueprint(api_2_0_blueprint, url_prefix='/api/v2.0')

    from .api_2_1 import api as api_2_1_blueprint
    app.register_blueprint(api_2_1_blueprint, url_prefix='/api/v2.1')
```

But this causes: `AssertionError: A blueprint's name collision occurred between <flask.blueprints.Blueprint object at 0x7f8e48e82c10> and <flask.blueprints.Blueprint object at 0x7f8e48ef7150>.  Both share the same name "api".  Blueprints that are created on the fly need unique names.`

It's true that both are called `api` inside their folders but I'm importing them with different names. Having to rename all the calls to `api` to something else for each version would make versioning painful and overall a code spaghetti.

Is there a better way to do this?

**Comment by RonnyPfannschmidt** [comment]
no, thats per design - blueprints express reusable parts of an application, each part has a different name

they are not quite intended for api versioning at all

**Comment by alejoar** [comment]
So is there a pain free way to do versioning with flask?

**Comment by ThiefMaster** [comment]
Why can't you rename the old blueprints? I guess you don't build urls for old api versions using `url_for` anyway, right?

**Comment by alejoar** [comment]
Ok so it turns out I only had to change the Blueprint definition.

Before I had both blueprints defined as `api = Blueprint('api', __name__)` which caused the collision and made me think that I needed to change name of the variable `api`. Turns out what I really needed to change is just the string `'api'` used in the call to `Blueprint`, so now my blueprints are defined as api = Blueprint('api_2_0', __name__) and `api = Blueprint('api_2_1', __name__)`, allowing me to keep the variable `api` the same in both cases and saving me the issue of chaging it everywhere.

I'm closing this issue now, sorry for the silly question.

**Comment by RonnyPfannschmidt** [comment]
@alejoar its not a silly question and we are most pleased the resolving your misconception was relatively painless

i wonder if some documentation additions might help future developers to understand directly

**Comment by jeffwidman** [comment]
This might also be helpful: http://stackoverflow.com/a/28797512/770425

**Comment by alejoar** [comment]
@jeffwidman thank you. I actually [asked a question](http://stackoverflow.com/questions/36474042/api-versioning-with-blueprints-causes-colission) in stackoverflow referencing that answer (I don't have enough rep to comment in other people's answers yet).

Luckily I already solved the issue. The problem is that I named both the variable and the blueprint the same and that misled me on where the collision was happening.
